src/single_random_draw.rs

```diff
@@ -0,0 +1,166 @@
+//! This library provides efficient algorithms to compose a set of unspent transaction outputs
+//! (UTXOs).
+
+#[cfg(any(test, feature = "rand"))]
+use crate::errors::LibError;
+
+use bitcoin::Amount;
+use bitcoin::blockdata::transaction;
+use bitcoin::FeeRate;
+use bitcoin::TxOut;
+use bitcoin::Weight;
+use crate::CHANGE_LOWER;
+use rand::seq::SliceRandom;
+
+/// Randomly select coins for the given target by shuffling the utxo pool and
+/// taking UTXOs until the given target is reached.
+#[cfg_attr(docsrs, doc(cfg(feature = "rand")))]
+pub fn select_coins_srd<R: rand::Rng + ?Sized>(
+    target: Amount,
+    outputs: &mut Vec<TxOut>,
+    fee_rate: FeeRate,
+    rng: &mut R)
+-> Result<Option<Vec<TxOut>>, LibError> {
+    outputs.shuffle(rng);
+
+    let mut selected_effective_value: Amount = Amount::ZERO;
+    let mut i = 0;
+    let mut result = Vec::new();
+
+    // The required size of a change output must be at least
+    // larger than the target by CHANGE_LOWER.  Therefore,
+    // CHANGE_LOWER + target becomes the new lower bound for the
+    // selection.
+    while selected_effective_value < target + CHANGE_LOWER && i < outputs.len() {
+        let output: &TxOut = &outputs[i];
+        let value: Amount = output.value;
+
+        let outputs = outputs.iter().map(|txout| txout.script_pubkey.len());
+
+        let weight: Weight = transaction::predict_weight(
+            None::<Vec<transaction::InputWeightPrediction>>,
+            Some(outputs),
+        );
+
+        let fee: Option<Amount> = fee_rate.checked_mul_by_weight(weight);
+
+        match fee {
+            Some(f) => {
+                let effective_value = value - f;
```

Comment:
the effective value seem to be computed using the fee for all outputs, which seems incorrect to me. I guess it should only use the fee for the added input?

Reply:
> Shouldn't it be checking the fees for inputs rather than outputs?

Well, based on your previous comment it should be the fee for all inputs?  Also, I think this will be corrected if the [line here](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/23/files/3ab168d23d2ab2c52bdda669ae6eb5fcd99df1de#diff-ce2c6b6b3132bf7b26a80fc104adfdeb0eb023b9bfd487c0da95d83830a73a32R38) is changed to be the selected "inputs".

Reply:
Agree with @Tibo-lg, the effective value of an input is the UTXO’s amount minus the cost of spending that UTXO at a given feerate. Also be sure to skip UTXOs with a negative effective value in selection.

Reply:
I've update this so the effective_value is now based on the sum of the utxos so far selected.

Reply:
> Also be sure to skip UTXOs with a negative effective value in selection.

Amount is of type U64 which can't be negative, so rust-bitcoin produces a panic if this happens.

Reply:
Then wouldn't it be better to do a `checked_sub`? It feels the `> 0` check is meaningless for now.

Reply:
Good idea, thanks! done.

Reply:
This will now skip any utxos with a negative effective value.